src/main/scala/analysis/GlobalRegionAnalysis.scala

```diff
@@ -149,35 +118,48 @@ trait GlobalRegionAnalysis(val program: Program,
    * @return Set[DataRegion]
    */
   def checkIfDefined(dataRegions: Set[DataRegion], n: CFGPosition): Set[DataRegion] = {
-    dataRegions.map { i =>
+    var converted: Set[DataRegion] = Set.empty
+    dataRegions.foreach { i =>
       val (f, p) = mmm.findDataObjectWithSize(i.start, i.size)
       val accesses = f.union(p)
       if (accesses.isEmpty) {
         i
       } else if (accesses.size == 1) {
-        dataMap(i.start) = DataRegion(i.regionIdentifier, i.start, i.size.max(accesses.head.size))
-        dataMap(i.start)
+        if (f.contains(accesses.head)) {
+          // full access
+          dataMap(i.start) = DataRegion(i.regionIdentifier, i.start, i.size.max(accesses.head.size))
+          converted = converted ++ Set(dataMap(i.start))
+        } else {
+          // partial access (we cannot determine the size)
+          dataMap(i.start) = DataRegion(i.regionIdentifier, i.start, i.size)
+          converted = converted ++ Set(dataMap(i.start))
+        }
       } else {
-        val highestRegion = accesses.maxBy(_.start)
-        dataMap(i.start) = DataRegion(i.regionIdentifier, i.start, i.size.max(highestRegion.end - i.start))
-        dataMap(i.start)
+  //        val highestRegion = accesses.maxBy(_.start)
+  //        dataMap(i.start) = DataRegion(i.regionIdentifier, i.start, i.size.max(highestRegion.end - i.start))
+  //        dataMap(i.start)
+        dataMap.remove(i.start)
+        accesses.foreach(a => dataMap(a.start) = dataPoolMaster(a.start, a.size))
+        converted = converted ++ accesses.collect({ case a => dataMap(a.start) })
       }
     }
+    converted
   }
 
   /** Transfer function for state lattice elements.
    */
   def localTransfer(n: CFGPosition, s: Set[DataRegion]): Set[DataRegion] = {
     n match {
       case memAssign: MemoryAssign =>
-        checkIfDefined(evalMemLoadToGlobal(memAssign.index, memAssign.size, memAssign), n)
+        return checkIfDefined(tryCoerceIntoData(memAssign.index, memAssign, memAssign.size), n)
       case assign: Assign =>
         val unwrapped = unwrapExpr(assign.rhs)
         if (unwrapped.isDefined) {
-          checkIfDefined(evalMemLoadToGlobal(unwrapped.get.index, unwrapped.get.size, assign, loadOp = true), n)
+          val regions: Set[DataRegion] = tryCoerceIntoData(unwrapped.get.index, assign, unwrapped.get.size)
+          return checkIfDefined(regions, n)
         } else {
           // this is a constant but we need to check if it is a data region
-          checkIfDefined(evalMemLoadToGlobal(assign.rhs, 1, assign), n)
+          return checkIfDefined(tryCoerceIntoData(assign.rhs, assign, 1), n)
```

Comment:
Why does this use an access size of 1? This seems to produce a lot of regions with size 1 in the GRA results that are not actually real regions.